<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
query_3_1 = f'''select count(id) "Количество вакансий"
                from public.vacancies
'''

In [5]:
df = pd.read_sql_query(query_3_1,connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/1742807029.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1,connection)


Количество вакансий
0                49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
query_3_2 = f'''select count(id) "Количество работодателей"
                from public.employers
'''

In [7]:
df = pd.read_sql_query(query_3_2, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/3588252056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


Количество работодателей
0                     23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
query_3_3 = f'''select count(id)  "Количество регионов"
                from public.areas
'''

In [9]:
df = pd.read_sql_query(query_3_3, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/2018224393.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


Количество регионов
0                 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
query_3_4 = f'''select count(id) "Количество сфер деятельности"
                from public.industries
'''

In [11]:
df = pd.read_sql_query(query_3_4, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/2172479966.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


Количество сфер деятельности
0                           294

***

***Выводы по предварительному анализу данных:***
- в среднем на одну сферу деятельности приходится по 167 вакансий;
- в среднем на каждый из регионов приходится по 36 вакансий;
- большое количество регионов, в которых зарегистрированы компании обусловлено тем, что компания HeadHunter развита не только на территории Российской федерации, но также и в Казахстане, Белоруси и других государствах. 
- в среднем каждая сфера деятельности представлена 80 работодателями и 167 вакансиями;

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
query_4_1 = f'''select 
                    count(v.id) "Количество вакансий", 
                    a.name "Регион"
                from 
                    public.vacancies v
                    left join public.areas a on v.area_id = a.id
                group by a.id
                order by 1 desc
'''

In [13]:
df = pd.read_sql_query(query_4_1, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


Количество вакансий                    Регион
0                   5333                    Москва
1                   2851           Санкт-Петербург
2                   2112                     Минск
3                   2006               Новосибирск
4                   1892                    Алматы
..                   ...                       ...
764                    1                    Кизляр
765                    1                    Джизак
766                    1                    Эртиль
767                    1                  Арсеньев
768                    1  Тура (Красноярский край)

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
#оператором WHERE осуществляем проверку нижней или верхней границ зарплатной вилки на отсутсвие пустых значений
query_4_2 = f'''select 
                    count(id) "Количество вакансий"
                from
                    public.vacancies
                where salary_from is not null or salary_to is not null 
'''

In [15]:
df = pd.read_sql_query(query_4_2, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/2168503709.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


Количество вакансий
0                24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
query_4_3 = f'''select 
                    round(avg(salary_from)) "Нижняя граница зарплатной вилки",
                    round(avg(salary_to)) "Верхняя граница зарплатной викли"
                from
                    public.vacancies
'''

In [18]:
df = pd.read_sql_query(query_4_3, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/2222372203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


Нижняя граница зарплатной вилки  Верхняя граница зарплатной викли
0                          71065.0                          110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
query_4_4 = f'''select 
                    schedule "Тип рабочего графика",
                    employment "Тип трудоустройства",
                    count(id) "Количество вакансий"
                from
                    public.vacancies
                group by
                    schedule,
                    employment
                order by 3 desc
'''

In [20]:
df = pd.read_sql_query(query_4_4, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


Тип рабочего графика  Тип трудоустройства  Количество вакансий
0           Полный день     Полная занятость                35367
1      Удаленная работа     Полная занятость                 7802
2         Гибкий график     Полная занятость                 1593
3      Удаленная работа  Частичная занятость                 1312
4        Сменный график     Полная занятость                  940
5           Полный день           Стажировка                  569
6        Вахтовый метод     Полная занятость                  367
7           Полный день  Частичная занятость                  347
8         Гибкий график  Частичная занятость                  312
9           Полный день     Проектная работа                  141
10     Удаленная работа     Проектная работа                  133
11        Гибкий график           Стажировка                  116
12       Сменный график  Частичная занятость                  101
13     Удаленная работа           Стажировка                   64
14        Гибкий график     Проектная работа                   18
15       Сменный график           Стажировка                   12
16       Вахтовый метод     Проектная работа                    2
17       Сменный график     Проектная работа                    1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
query_4_5 = f'''select 
                    experience as "Требуемый опыт работы",
                    count(id) "Количество вакансий"
                from
                    public.vacancies
                group by experience
                order by 2
'''

In [22]:
df = pd.read_sql_query(query_4_5, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/2601166264.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


Требуемый опыт работы  Количество вакансий
0           Более 6 лет                 1337
1             Нет опыта                 7197
2         От 3 до 6 лет                14511
3    От 1 года до 3 лет                26152

Дополнительное исследование 4.1. Напишите запрос, выводящий значений нижней и верхней границы зарплатной вилки, название вакансии и название города, где размещена вакансия. Вывести только те вакансии где указана нижняя зарплатная вилка. Сортировку произвести по нижней границе зарплатной вилки.

In [23]:
query_add_res_4_1 = f'''select 
                    v.salary_from "Нижняя граница зарплатной вилки", 
                    v.salary_to "Верхняя граница зарплатной вилки", 
                    v.name "Назкание вакансии", 
                    a.name "Название города"
                from
                    public.vacancies v
                    left join public.areas a on v.area_id = a.id
                where v.salary_from is not null
                order by 1
'''

In [24]:
df_salary_from = pd.read_sql_query(query_add_res_4_1, connection)
df_salary_from

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/1513858537.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_salary_from = pd.read_sql_query(query_add_res_4_1, connection)


Нижняя граница зарплатной вилки  Верхняя граница зарплатной вилки  \
0                                   30                           50000.0   
1                                   30                               NaN   
2                                   30                               NaN   
3                                   35                               NaN   
4                                   42                           50000.0   
...                                ...                               ...   
21480                           500000                               NaN   
21481                           574000                          860000.0   
21482                           574000                          860000.0   
21483                           650000                               NaN   
21484                          1000000                               NaN   

                                       Назкание вакансии  \
0                                Системный администратор   
1                                Помощник IT-специалиста   
2                                Помощник IT-специалиста   
3                 Инженер по информационной безопасности   
4      Инженер по обслуживанию ресторанов Макдоналдс ...   
...                                                  ...   
21480  Директор научно-исследовательского центра (химия)   
21481              Frontend Developer in Unicorn Company   
21482             FullStack Developer in Unicorn Company   
21483             Директор по информационным технологиям   
21484                               Технический директор   

                    Название города  
0                         Челябинск  
1                              Сочи  
2                              Сочи  
3                           Воронеж  
4                           Воронеж  
...                             ...  
21480  Иваново (Ивановская область)  
21481                        Алматы  
21482                        Алматы  
21483                        Москва  
21484               Санкт-Петербург  

[21485 rows x 4 columns]

Дополнительное исследование 4.2. Напишите запрос, выводящий значений нижней и верхней, название вакансии и название города, где размещена вакансия. Вывести только те вакансии где указана верхняя зарплатная вилка. Сортировку произвести по врехней границе зарплатной вилки по убыванию.

In [25]:
query_add_res_4_2 = f'''select 
                    v.salary_from "Нижняя граница зарплатной вилки", 
                    v.salary_to "Верхняя граница зарплатной вилки", 
                    v.name "Назкание вакансии", 
                    a.name "Название города"
                from
                    public.vacancies v
                    left join public.areas a on v.area_id = a.id
                where v.salary_to is not null
                order by 2 desc
'''

In [26]:
df_salary_to = pd.read_sql_query(query_add_res_4_2, connection)
df_salary_to

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/1051209122.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_salary_to = pd.read_sql_query(query_add_res_4_2, connection)


Нижняя граница зарплатной вилки  Верхняя граница зарплатной вилки  \
0                             400000.0                           1000000   
1                             500000.0                           1000000   
2                             500000.0                           1000000   
3                             350000.0                            950000   
4                             175346.0                            935176   
...                                ...                               ...   
12685                           1500.0                              1500   
12686                            750.0                               850   
12687                            250.0                               350   
12688                              NaN                                75   
12689                             60.0                                60   

                                       Назкание вакансии  Название города  
0                                 QA Automation Engineer           Алматы  
1                             СТО/Директор по разработке           Москва  
2                         Брокер по продаже апартаментов             Сочи  
3                    1С:Эксперт по производительности 1С           Москва  
4      Senior Software Developer (upwork / иностранны...  Санкт-Петербург  
...                                                  ...              ...  
12685  Подработка на выходные разнорабочий на станцию...         Кемерово  
12686             Консультант 1С - бухгалтерия, зарплата        Хабаровск  
12687                           Computer Vision Engineer           Москва  
12688                                        Программист         Владимир  
12689  Database Administrator (Администратор баз данных)         Смоленск  

[12690 rows x 4 columns]

***

In [27]:
import plotly
import plotly.express as px

fig = px.box(
    data_frame=df_salary_from,
    x = 'Нижняя граница зарплатной вилки',
    title= 'Коробчатая диаграмма анализа нижней границы зарплатной вилки'
)
fig.show()

In [28]:

fig = px.box(
    data_frame=df_salary_to,
    x = 'Верхняя граница зарплатной вилки',
    title= 'Коробчатая диаграмма анализа верхней границы зарплатной вилки'
)
fig.show()

***Выводы по детальному анализу вакансий:***
- Ожидаемо, большинство вакансий представлены в мегаполисах с населением более 1 миллиона человек. ТОП-5 регионов представляют 14194 вакансий, что составляет 29% от общего числа вакансий;
- Заполняемость графы с указанием зарплаты в вакансиях составляет 48.9%, что затруднит анализ по определению уровня зарплаты. В дальнейшем в фазе DataPreparation необходимо решить данную проблему либо удаление пустых значений либо их заполнением.
- Средний нижний уровень зарплатной вилки составляет 71065 рублей, что на 13% выше среднего уровня зарплат по России (данные [данные Росстата](https://gogov.ru/articles/average-salary) за ноябрь 2022 года - 63060р.). Это может быть связано с тем, что компания HeadHunter является крупнейшей российской интернет-рекрутмент компанией. Доступ к базе резюме и размещение вакансий платные, в связи с чем не все работодатели видят смысл или имеют возможность работать с этой платформой (особенно это касается вакансий не требующих какой либо квалификации или низкооплачиваемых вакансий);
- В разрезе сочетания типов рабочего графика и типа трудоустройства наиболее предпочтительным является полный день с полной занятостью: 35367 вакансий(71.8% от общего числа вакансий), а также удаленная работа с полной занятостью: 7802 вакансии (15%). Полный день с полной занятостью является наиболее предпочтительным так как: ведется наиболее рациональная организация и координация работы всех структур и подразделений, имеется возможность оперативного контроля соблюдения графика выхода на работу, удобнее вести учет и нормирование труда, наличие обширной клиентской базы соискателей для трудоустройства. В то же время в связи с переводом многих сфер деятельности в онлайн быстро набирает популярность удаленная работа с полной занятостью, которая является более удобной как для работодателя (отсутствие необходимости организации рабочего пространства для персонала, снятие проблем с трансфером персонала), так и для соискателей (отсутствие материальных и временных затрат на трансфер, собственная рабочая обстановке, свободный дресс-код, нет привязки к определенному местоположению, гибкий рабочий график, больше времени уделяется семье);
- Наиболее распространенный требуемый опыт работы в вакансиях: от 1 года до 3 лет (53%) и от 3 до 6 лет (29%). Наименее востребованным уровнями опыта работы являются более 6 лет (3%) и без опыта (15%). Это свидетельствует о наибольшей востребованности молодых специалистов с базовым опытом работы. Данная категорию людей наиболее быстро приспосабливается к новым условиям работы и легче обучаема. В то же время наибольшие трудности с поиском работы возникают у специалистов без опыта работы. Для оценки данной проблемы необходимо провести дополнительный анализ вакансий для людей без опыта работы с целью определения наиболее подходящего типа рабочего графика и типа трудоустройства для них.
- При расширенном анализе нижней и верхней границ зарплатной вилки выявляются предположительно аномальные значения, такие как: минимальный уровень зарплаты 30 р/месяц (при установлено МРОТ РФ на 2023 год в размере 16242р.), одинаковые показатели зарплат для различных вакансий в других государствах, максимальный уровень зарплаты в 1000000р, не характерный для вакансии QA Automation Engineer. Предположения об аномальных значениях подтверждаются коробчатыми диаграммами. Данный факт необходимо учесть при дальнейшей очистке данных.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [29]:
#две строки вывода формируем с помощью оператора UNION ALL с написанием отдельных запросов для каждой из строк
query_5_1 = f'''(select 
                    e.name "Название работодателя",
                    count(v.id) "Количество вакансий"
                from
                    public.vacancies v 
                    left join public.employers e on v.employer_id = e.id
                group by e.id
                order by 2 desc
                limit 1)
                union all
                (select 
                    e.name,
                    count(v.id)
                from
                    public.vacancies v 
                    left join public.employers e on v.employer_id = e.id
                group by e.id
                order by 2 desc
                offset 4 limit 1)        
'''

In [30]:
df = pd.read_sql_query(query_5_1, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/1792814825.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название работодателя  Количество вакансий
0                Яндекс                 1933
1         Газпром нефть                  331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [31]:
query_5_2 = f'''select 
                    a.name "Назкание города",
                    count(e.id) "Количество работодателей",
                    count(v.id) "Количество вакансйи"
                from
                    public.areas a
                    left join public.vacancies v on v.area_id = a.id
                    left join public.employers e on e.area = a.id
                group by a.id
                order by 3, 2 desc
'''

In [32]:
df = pd.read_sql_query(query_5_2, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/3341481329.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Назкание города  Количество работодателей  Количество вакансйи
0                 Россия                       410                    0
1              Казахстан                       207                    0
2     Московская область                        75                    0
3     Краснодарский край                        19                    0
4               Беларусь                        18                    0
...                  ...                       ...                  ...
1357         Новосибирск                   1149438              1149438
1358              Алматы                   1364132              1364132
1359               Минск                   2354880              2354880
1360     Санкт-Петербург                   6320667              6320667
1361              Москва                  31272712             31272712

[1362 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [33]:
query_5_3 = f'''select 
                    e.name "Название работодателя",
                    count(distinct v.area_id)  "Количество регионов"
                from
                    public.vacancies v
                    left join public.employers e on v.employer_id = e.id
                group by e.id
                order by 2 desc
'''

In [34]:
df = pd.read_sql_query(query_5_3, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/2398953733.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название работодателя  Количество регионов
0                                      Яндекс                  181
1                                  Ростелеком                  152
2                                  Спецремонт                  116
3                      Поляков Денис Иванович                   88
4                                   ООО ЕФИН                    71
...                                       ...                  ...
14901                            НПП Авиатрон                    1
14902              Центр дистанционных торгов                    1
14903  Городские Телекоммуникационные Системы                    1
14904                       Введенский, Отель                    1
14905              Группа Компаний Автоимпорт                    1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [35]:
query_5_4 = f'''select 
                    count(e.id)  "Количество работодателей"
                from
                    public.employers e 
                    left join public.employers_industries ei on e.id = ei.employer_id
                where ei.employer_id is null
'''

In [36]:
df = pd.read_sql_query(query_5_4, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/3114411492.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество работодателей
0                      8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [37]:
query_5_5 = f'''select 
                    e.name "Название работодателя"
                from
                    public.employers e 
                    left join public.employers_industries ei on e.id = ei.employer_id
                group by e.id
                having count(ei.industry_id) = 4
                order by 1
                offset 2 limit 1
'''

In [38]:
df = pd.read_sql_query(query_5_5, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/1650419958.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название работодателя
0                  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [39]:
query_5_6 = f'''select 
                    count(e.id) "Количество работодателей"
                from
                    public.employers e 
                    left join public.employers_industries ei on e.id = ei.employer_id
                    left join public.industries i on ei.industry_id = i.id
                where i.name like 'Разработка программного обеспечения'
'''

In [40]:
df = pd.read_sql_query(query_5_6, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/1169039133.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество работодателей
0                      3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [41]:
# Возникшую ошибку SSLCertVerificationError устраняем с помошью импрота модуля SSL и заданием нового контекста с безопасными настройками по умолчанию
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

# Записываем URL старницы в Википедии с указанием городов-милионников
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8#%D0%A1%D0%BF%D0%B8%D1%81%D0%BE%D0%BA_%D0%B3%D0%BE%D1%80%D0%BE%D0%B4%D0%BE%D0%B2-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D0%BE%D0%B2'
tables = pd.read_html(url) # записываем в переменную HTML-таблицы с указанной страницы
city_df = tables[1]  # записываем в переменную таблицу с указанием городов - миллионников
city_list = list(city_df['Город']) # создаем список городов-миллионников
city_list[13] = city_list[13][:-3] # вручную корректируем записо одного из городов, записаного с дополнением
city_list = tuple(city_list) # переводим список в кортеж
print(city_list)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')


In [42]:
# создем запрос, в котором фильтрацию осуществляем путем передачи сохраненного ранее кортежа с помощью f-строки в наш запрос
query_5_7 = f'''(select 
                    a.name "Название города",
                    count(v.id) "Количество вакансий"
                from
                    public.vacancies v
                    left join employers e on v.employer_id = e.id
                    left join areas a on v.area_id = a.id
                where a.name in {city_list} and e.name = 'Яндекс'
                group by a.name)
                union all
                (select 
                    'Total',
                    count(v.id)
                from
                    public.vacancies v
                    left join employers e on v.employer_id = e.id
                    left join areas a on v.area_id = a.id
                where a.name in {city_list} and e.name = 'Яндекс')
                order by 2
'''

In [43]:
df = pd.read_sql_query(query_5_7, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/3129374555.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название города  Количество вакансий
0              Омск                   21
1         Челябинск                   22
2        Красноярск                   23
3         Волгоград                   24
4             Пермь                   25
5            Казань                   25
6    Ростов-на-Дону                   25
7               Уфа                   26
8            Самара                   26
9         Краснодар                   30
10          Воронеж                   32
11      Новосибирск                   35
12  Нижний Новгород                   36
13     Екатеринбург                   39
14  Санкт-Петербург                   42
15           Москва                   54
16            Total                  485

***

***Выводы по анализу работодателей:***
- Наиболее активный рекрутинг показывает компания Яндекс: 1933 вакансии(3.9% от общего числа вакансий), что свидетельствует о высокм уровне развития и расширения данной компании.
- В данных представлены компании, зарегистрированные не в конкретном городе, а в области или республике, но вакансии данных кампаний размещены в конкретных населенных пунктах.
- Наиболее широко распространенной (в разрезе задействованных регионов) компанией ожидаемо является Яндекс (181 регион или 13.2%). Это также подтверждает сведения о высоком уровне развития компании. 
- У 8419 компаний (35.8% от общего числа работодателей) отсутствуют сведения о сфере деятельности, что может осложнить подбор специалистов для данных компаний.
- В качестве сферы деятельности "Разработка программного обеспечения" заполнили 3553 компаний (15% от общего числа работодателей и 55.8% от числа работодателей, заполнивших раздел о сфере деятельности) что свидетельствует о востребованности специалистов IT-профиля.
- Компания Яндекс проводит наиболее массовый рекрутинг специалистов городах с населением более 1 миллиона человек (от 21 до 54 вакансий на каждый город). Суммарное количество вакансий для городов миллионников от данной компании составляет 485 (25% от общего числа вакансий данной компании).  

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [44]:
query_6_1 = f'''select 
                    count(id) "Количество вакансий"
                from
                    public.vacancies
                where lower(name) like '%data%' or lower(name) like '%данн%'
'''

In [45]:
df = pd.read_sql_query(query_6_1, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/485020324.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество вакансий
0                 1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [46]:
# при фильтрации используем опреатро ILIKE, чтобы исключить необходимость перевода данных к единому регистру
query_6_2 = f'''select 
                    count(id) "Количество вакансий"
                from
                    public.vacancies
                where   (name ilike '%data scientist%' or
                        name ilike '%data science%' or 
                        name ilike '%исследователь данных%' or
                        (name LIKE '%ML%' AND name NOT LIKE '%HTML%') or
                        name ilike '%machine learning%' or
                        name ilike '%машинн%обучен%')
                        and
                        (name ilike '%junior%' or
                        experience ilike '%нет опыта%' or
                        employment ilike '%стажировка%')
'''

In [47]:
df = pd.read_sql_query(query_6_2, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/3641650667.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество вакансий
0                   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [48]:
query_6_3 = f'''select 
                    count(id) "Количество вакансий"
                from
                    public.vacancies
                where   (name ilike '%data scientist%' or
                        name ilike '%data science%' or 
                        name ilike '%исследователь данных%' or
                        (name LIKE '%ML%' AND name NOT LIKE '%HTML%') or
                        name ilike '%machine learning%' or
                        name ilike '%машинн%обучен%')
                        and
                        (key_skills like '%SQL%' or
                        key_skills ilike '%postgres%')
'''

In [49]:
df = pd.read_sql_query(query_6_3, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/1159587760.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество вакансий
0                  201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [50]:
query_6_4 = f'''select 
                    count(id) "Количество вакансий"
                from
                    public.vacancies
                where   (name ilike '%data scientist%' or
                        name ilike '%data science%' or 
                        name ilike '%исследователь данных%' or
                        (name LIKE '%ML%' AND name NOT LIKE '%HTML%') or
                        name ilike '%machine learning%' or
                        name ilike '%машинн%обучен%')
                        and
                        key_skills like '%Python%'
'''

In [51]:
df = pd.read_sql_query(query_6_4, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/972386973.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество вакансий
0                  351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [52]:
query_6_5 = f'''select 
                    round(avg(length(key_skills) - length(replace(key_skills, CHR(9),''))+1),2) "Среднее количество навыков"
                from
                    public.vacancies
                where   (name ilike '%data scientist%' or
                        name ilike '%data science%' or 
                        name ilike '%исследователь данных%' or
                        (name LIKE '%ML%' AND name NOT LIKE '%HTML%') or
                        name ilike '%machine learning%' or
                        name ilike '%машинн%обучен%')
                        and 
                        key_skills is not null
'''

In [53]:
df = pd.read_sql_query(query_6_5, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/390248827.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Среднее количество навыков
0                        6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [54]:
# Функция COALESCE возвращает первый попавшийся аргумент, отличный от NULL
query_6_6 = f'''select 
                    experience "Требования к опыту",
                    round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to))) "Средня зарплата"
                from
                    public.vacancies
                where   (name ilike '%data scientist%' or
                        name ilike '%data science%' or 
                        name ilike '%исследователь данных%' or
                        (name LIKE '%ML%' AND name NOT LIKE '%HTML%') or
                        name ilike '%machine learning%' or
                        name ilike '%машинн%обучен%')
                        and 
                        (salary_from is not null or
                        salary_to is not null)
                        and
                        experience = 'От 3 до 6 лет'
                group by experience
'''

In [55]:
df = pd.read_sql_query(query_6_6, connection)
df

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/3561045109.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Требования к опыту  Средня зарплата
0      От 3 до 6 лет         243115.0

Дополнительное исследование 6.1. Напишите запрос, выводящий ТОП-10 компаний, разместивших вакансии на позицию DataScientist с указанием количества вакансий. Сортировку прооизвести по убыванию количества вакансий.

In [56]:
query_add_res_6_1 = f'''select 
                    e.name "Название работодателя", 
                    count(v.id) "Количество вакансий"
                from
                    public.vacancies v
                    left join public.employers e on v.employer_id = e.id
                where   (v.name ilike '%data scientist%' or
                        v.name ilike '%data science%' or 
                        v.name ilike '%исследователь данных%' or
                        (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%') or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                group by e.id
                order by 2 desc
                limit 10
'''

In [57]:
df_salary_to = pd.read_sql_query(query_add_res_6_1, connection)
df_salary_to

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/538793153.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название работодателя  Количество вакансий
0                   СБЕР                   37
1        Bell Integrator                   25
2         Банк ВТБ (ПАО)                   18
3                     VK                   15
4  Positive Technologies                   11
5         EvenBet Gaming                    9
6                 Яндекс                    9
7                МегаФон                    8
8               Andersen                    7
9                    МТС                    6

Дополнительное исследование 6.2. Напишите запрос, выводящий ТОП-10 компаний, разместивших вакансии на позицию DataScientist уровня Junior с указанием количества вакансий. Сортировку прооизвести по убыванию количества вакансий.

In [58]:
query_add_res_6_2 = f'''select 
                    e.name "Название работодателя", 
                    count(v.id) "Количество вакансий"
                from
                    public.vacancies v
                    left join public.employers e on v.employer_id = e.id
                where   (v.name ilike '%data scientist%' or
                        v.name ilike '%data science%' or 
                        v.name ilike '%исследователь данных%' or
                        (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%') or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                        and
                        (v.name ilike '%junior%' or
                        v.experience ilike '%нет опыта%' or
                        v.employment ilike '%стажировка%')
                group by e.id
                order by 2 desc
                limit 5
'''

In [59]:
df_salary_to = pd.read_sql_query(query_add_res_6_2, connection)
df_salary_to

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/2832231502.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название работодателя  Количество вакансий
0                                    Банк ВТБ (ПАО)                   10
1  The Skolkovo Institute of Science and Technology                    3
2                                            Comexp                    3
3     СтарЛайн, Научно-производственное объединение                    2
4                                     SteadyControl                    2

***

Дополнительное исследование 6.3. Напишите запрос, выводящий ТОП-5 городов с максимальным количеством вакансий для специалистов DataScience. Сортировку произвести по убыванию количества вакансий

In [60]:
query_add_res_6_3 = f'''select 
                    a.name "Название города", 
                    count(v.id) "Количество вакансий"
                from
                    public.vacancies v
                    left join public.areas a on v.area_id = a.id
                where   (v.name ilike '%data scientist%' or
                        v.name ilike '%data science%' or 
                        v.name ilike '%исследователь данных%' or
                        (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%') or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                group by a.id
                order by 2 desc
                limit 10
'''

In [61]:
df_salary_to = pd.read_sql_query(query_add_res_6_3, connection)
df_salary_to

/var/folders/wc/hjmp0jxd4s59kh06mhqwbjlc0000gn/T/ipykernel_961/3763708262.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название города  Количество вакансий
0           Москва                  217
1  Санкт-Петербург                   64
2      Новосибирск                   23
3  Нижний Новгород                   20
4           Алматы                   16
5           Казань                   16
6            Минск                    9
7            Томск                    8
8     Екатеринбург                    7
9          Армения                    6

***Выводы по предметному анализу:***
- Всего вакансий, имеющих отношение к данным выставлено 1771 (49,8% от числа вакансий, связанных с разработкой программного обеспечения), что свидетельствует о высоком уровне востребованности специалистов по работе с базами данных в сфере IT и их высокими темпами распространения.
- Отмечается низкий уровень востребованности начинающих специалистов DataScience (51 вакансия или 2.8%), что требует от данных специалистов повышения уровня квалификации на различных курсах или путем приобретения опыта на стажировочных вакансиях.
- Навык пользования SQL и Postress отмечен в требования к 201 вакансии (11,3% от вакансий DataScience) что свидетельствует о важности и востребованности данных знаний у кандидатов, так же как и знания Python - 351 вакансия(19,8%).
- Специалист dataScience должен обладать в среднем 6-7 ключевыми навыками для трудоустройства по данной специальности
- Средний уровень зарплаты специалиста DataScience без опыта работы составляет 74643 рубля, что выше общего среднего нижнего уровня зарплат, а разница средней зарплаты между специалистом без опыта, специалистом с опытом от 1 до 3 лет, и специалистом с опытом более 6 лет составляет 65032 рубля (выше на 87%) и 168472 рубля (выше на 126%) соответственно. Данный факт указывает на ценность специалистов с богатым опытом работы;
- Компании, наиболее активно осуществляющие рекрутинг специалистов в сфере DataScience, являются СБЕР, Bell Integrator, Банк ВТБ (ПАО), VK, Positive Technologies. При этом компания Банк ВТБ (ПАО) активно рекрутирует специалистов DataScience уровня Junior (10 вакансий из 18);
- Города, где наиболее востребованы специалисты в сфере DataScience, ожидаемо являются крупные города с населением более 1 миллиона человек;

# Общий вывод по проекту

- Объем и качество данных позволяют производить дальнейшую работу с данными.
- На этапе Data Preparation необходимо обратить внимание на наличие предполагаемых выбросов и аномальных значений, а также на большое количество пропусков в признаке нижней и верхней границ зарплатной вилки по вакансиям;
- Также при моделировании необходимо учитывать больше количество пропусков в сферах деятельности работодателей;